server/src-lib/Hasura/GraphQL/Execute/LiveQuery/State.hs

```diff
@@ -107,16 +116,22 @@ addLiveQuery lqState plan onResultAction = do
     newPoller = Poller <$> TMap.new <*> STM.newEmptyTMVar
 
 removeLiveQuery
-  :: LiveQueriesState
+  :: L.Logger L.Hasura
+  -> LiveQueriesState
   -- the query and the associated operation
   -> LiveQueryId
   -> IO ()
-removeLiveQuery lqState (LiveQueryId handlerId cohortId sinkId) = do
-  threadRef <- STM.atomically $ do
+removeLiveQuery logger lqState lqId@(LiveQueryId handlerId cohortId sinkId) = do
+  threadRefM <- STM.atomically $ do
     detM <- getQueryDet
     fmap join $ forM detM $ \(Poller cohorts ioState, cohort) ->
       cleanHandlerC cohorts ioState cohort
-  traverse_ A.cancel threadRef
+  case threadRefM of
+    Just threadRef -> Immortal.stop threadRef
+    -- This would seem to imply addLiveQuery broke, or a bug elsewhere:
+    Nothing -> L.unLogger logger $ L.UnstructuredLog L.LevelError $ fromString $
```

Comment:
`Nothing` could also mean several things:
1. There is no associated live query with the specified `lqId`.
2. The given live query is not the last live query of the handler (a handler can manage more than one live query). 
3. If `fmap _pThread <$> STM.tryReadTMVar ioState` returns a `Nothing`. This is the case where it is definitely a bug, but what I do not understand (or recall) is why is `tryReadTMVar` being used instead of `readTMVar`?

So if we get rid of the 3rd case, this log isn't needed as `Nothing` is a valid value?

Reply:
Good catch, thanks! There are some comments elsewhere that explains the awkwardness behind this var. I think it's a bit safer to log than to use `readTMVar` and risk blocking in the case of a bug we didn't anticipate. Would be great if you had another look at the new version when I push it

Reply:
This is incorrect: 
```haskell
            case threadRefM of
              Just threadRef -> Immortal.stop threadRef
              -- This would seem to imply addLiveQuery broke or a bug
              -- elsewhere. Be paranoid and log:
              Nothing -> L.unLogger logger $ L.UnstructuredLog L.LevelError $ fromString $
                "In removeLiveQuery no worker thread installed. Please report this as a bug: "<>
                show lqId
```
You can have `Nothing` in any of the 3 scenarios mentioned above but only in (3) is when we need to log this message (maybe even (1) too because we wouldn't be calling removeLiveQuery on a query that cannot exist) but (2) is a valid scenario where `Nothing` is expected.

Reply:
@0x777 hm were you looking at an old revision? I believe the revision I made after your first comment addresses the issue (we only log a warning in case (3))

Reply:
@jberryman My bad!